### PR TITLE
Added the ReadmePreviewViewModel & edited NuGetPackageFileService

### PR DESCRIPTION
**Pull request by jgonz120**
… readme

# Bug

Part of: https://github.com/NuGet/Home/issues/12583

## Description
Adds a new ViewModel for getting the README contents, leveraging the NuGetPackageFileService to get the file from the package. 

## PR Checklist

- [X] Meaningful title, helpful description and a linked NuGet/Home issue
- [X] Added tests
- [X] ~~Link to an issue or pull request to update docs if this PR changes settings, environment variables, new feature, etc.~~

### CODE REVIEW

**Comment by jgonz120** [comment]
@nkolev92 this is going to a feature branch so I don't think there's a worry about merging.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/ViewModels/ReadmePreviewViewModelTests.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+using NuGet.VisualStudio.Internal.Contracts;
+using Xunit;
+
+namespace NuGet.PackageManagement.UI.Test.ViewModels
+{
+    [Collection(MockedVS.Collection)]
+    public class ReadmePreviewViewModelTests
+    {
+        [Fact]
+        public void ReadmePreviewViewModel_WithNullPackageFileService_ThrowsArgumentNullException()
```
Comment (jebriede) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/ViewModels/ReadmePreviewViewModelTests.cs]: Or Ctor

```suggestion
        public void Constructor_WithNullPackageFileService_ThrowsArgumentNullException()
```

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/ViewModels/ReadmePreviewViewModelTests.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            Assert.False(target.ErrorLoadingReadme);
+            Assert.Equal(string.Empty, target.ReadmeMarkdown);
+            Assert.True(target.CanDetermineReadmeDefined);
+        }
+
+        [Theory]
+        [InlineData(null)]
+        [InlineData("")]
+        [InlineData(" ")]
+        public async Task ReadmePreviewViewModel_LoadReadmeAsync_WithoutReadmeUrl_NoReadmeReturned(string readmeUrl)
```
Comment (jebriede) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/ViewModels/ReadmePreviewViewModelTests.cs]: We can leave the ReadmePreviewViewModel part off all the tests since this test class already describes the class being tested. See https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md#unit-test-method-naming

```suggestion
        public async Task LoadReadmeAsync_WithoutReadmeUrl_NoReadmeReturned(string readmeUrl)
```

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+        public async Task LoadReadmeAsync(string rawReadmeUrl, CancellationToken cancellationToken)
+        {
+            var newReadmeValue = string.Empty;
+            var isErrorWithReadme = false;
+            bool canDetermineReadmeDefined = false;
+
+            if (!string.IsNullOrWhiteSpace(rawReadmeUrl))
+            {
+
```
Comment (jebriede) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs:56]: nit: remove blank line.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        public bool CanDetermineReadmeDefined
+        {
+            get => _canDetermineReadmeDefined;
+            set
+            {
+                SetAndRaisePropertyChanged(ref _canDetermineReadmeDefined, value);
+            }
+        }
+
+        public async Task LoadReadmeAsync(string rawReadmeUrl, CancellationToken cancellationToken)
```
Comment (jebriede) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs:46]: Question: This isn't being used yet, except for in tests, correct?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                    using StreamReader streamReader = new StreamReader(readmeStream);
+                    var readme = await streamReader.ReadToEndAsync();
+                    if (!string.IsNullOrWhiteSpace(readme))
+                    {
+                        isErrorWithReadme = false;
+                        canDetermineReadmeDefined = true;
+                        newReadmeValue = readme;
+                    }
+                }
+                await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```
Comment (jebriede) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs]: We shouldn't switch to the main thread at the end of this method.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+        private bool _errorLoadingReadme;
+
+        public bool ErrorLoadingReadme
+        {
+            get => _errorLoadingReadme;
+            set
+            {
+                SetAndRaisePropertyChanged(ref _errorLoadingReadme, value);
+            }
```
Comment (jebriede) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs]: Style suggestion: More concise

```suggestion
            set => SetAndRaisePropertyChanged(ref _errorLoadingReadme, value);
```

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        public ReadmePreviewViewModel(INuGetPackageFileService packageFileService)
+        {
+            _packageFileService = packageFileService ?? throw new ArgumentNullException(nameof(packageFileService));
+            _errorLoadingReadme = false;
+            _canDetermineReadmeDefined = true;
+            _rawReadme = string.Empty;
+        }
+
+        private bool _errorLoadingReadme;
+
```
Comment (jebriede) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs]: Nit: remove blank line.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        {
+            get => _errorLoadingReadme;
+            set
+            {
+                SetAndRaisePropertyChanged(ref _errorLoadingReadme, value);
+            }
+        }
+
+        private string _rawReadme;
+
```
Comment (jebriede) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs]: Nit: remove blank line.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        {
+            get => _rawReadme;
+            set
+            {
+                SetAndRaisePropertyChanged(ref _rawReadme, value);
+            }
+        }
+
+        private bool _canDetermineReadmeDefined;
+
```
Comment (jebriede) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs]: Nit: remove blank line.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        public async Task LoadReadmeAsync(string rawReadmeUrl, CancellationToken cancellationToken)
+        {
+            var newReadmeValue = string.Empty;
+            var isErrorWithReadme = false;
+            bool canDetermineReadmeDefined = false;
+
+            if (!string.IsNullOrWhiteSpace(rawReadmeUrl))
+            {
```
Comment (jebriede) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs]: To reduce nested ifs we can take basic input validation checks like this and invert it to if (string.IsNullOrWhiteSpace(rawReadmeUrl)), set some defaults, and return.

**Review by jebriede — CHANGES_REQUESTED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs (hunk 4 above)
Comment (jgonz120) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs:46]: Yes, the next PR will be the UX where it'll be used

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+        public ReadmePreviewViewModel(INuGetPackageFileService packageFileService)
+        {
+            _packageFileService = packageFileService ?? throw new ArgumentNullException(nameof(packageFileService));
+            _errorLoadingReadme = false;
+            _canDetermineReadmeDefined = true;
+            _rawReadme = string.Empty;
+        }
+
+        private bool _errorLoadingReadme;
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs]: Our coding guidelines suggest separating groups of fields from properties.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs (hunk 5 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs]: I don't think updating properties on a background thread will work with WPF.
Jonatan, what happens if you remove the switch to main thread? I'm guessing either no UI update or an exception?

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageFileService.cs [hunk 12]
```diff
@@ -131,6 +131,23 @@ public NuGetPackageFileService(IServiceBroker serviceBroker, INuGetTelemetryProv
             return stream;
         }
 
+        public async ValueTask<Stream?> GetReadmeAsync(Uri readmeUri, CancellationToken cancellationToken)
+        {
+            Assumes.NotNull(readmeUri);
+
+            Stream? stream;
+            if (IsEmbeddedUri(readmeUri))
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageFileService.cs:172]: I recently changed the Icon algorithm to check for a local file first, then a remote file.
Do you think this is a good opportunity to do that for README? That is, check for the md file in GPF first, and only if not found, go to the remote?

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Services/NuGetPackageFileServiceTests.cs [hunk 13]
```diff
@@ -337,6 +337,37 @@ public async Task AddLicenseToCache_WhenMissing_IsNotFound()
...
+                var packageFileService = new NuGetPackageFileService(
+                        default(ServiceActivationOptions),
+                        Mock.Of<IServiceBroker>(),
+                        new AuthorizationServiceClient(Mock.Of<IAuthorizationService>()),
+                        _telemetryProvider.Object);
+
+                Stream readmeStream = await packageFileService.GetReadmeAsync(builder.Uri, CancellationToken.None);
+
+                Assert.NotNull(readmeStream);
+                Assert.Equal(StreamContents.Length, readmeStream.Length);
```
Comment (donnie-msft) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Services/NuGetPackageFileServiceTests.cs]: nit: it bothers me that the file content states it is an image, even though this is a readme 😆

**Review by donnie-msft — APPROVED** [review]
Thanks for making smaller PRs for this effort! I would reword the title since it also updates the file service and not just the VM. Approving with suggestions.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs (hunk 5 above)
Comment (jgonz120) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs]: From what I remember I got an exception when I tried to set the view model properties without switching to the main thread. I can remove it and we can see what happens in the next pr, it looks like the unit test passes either way

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Services/NuGetPackageFileServiceTests.cs (hunk 13 above)
Comment (jgonz120) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Services/NuGetPackageFileServiceTests.cs]: Did you mean the comments in the CreateDummyPackage? 😅

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageFileService.cs (hunk 12 above)
Comment (jgonz120) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageFileService.cs:172]: I don't think that will work for the Readme, I want to decide which source it's being loaded from. We don't want to display README in the browse tab if they aren't available from the remote source.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs (hunk 5 above)
Comment (jebriede) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs]: A better way to handle this is to use the [Dispatcher](https://learn.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatcher?view=windowsdesktop-8.0) to check if we're on the main thread or to marshal to the main thread.

Explicitly switching to the main thread is expensive and error-prone because it forces this background operation onto the main thread where it will remain, even when returning to the caller.

See: https://stackoverflow.com/questions/590590/making-%20sure-onpropertychanged-is-called-on-ui-thread-in-mvvm-wpf-app

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Services/NuGetPackageFileServiceTests.cs (hunk 13 above)
Comment (donnie-msft) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Services/NuGetPackageFileServiceTests.cs]: No, I mean StreamContents which is "hi I'm an image" or something.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageFileService.cs (hunk 12 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageFileService.cs:172]: I quickly re-read the draft spec and didn't see anything obvious indicating why READMEs should only be downloaded and not read from disk. https://github.com/NuGet/Home/pull/13200/files?short_path=6c7eb8e#diff-6c7eb8e1ec6b63a9782e440a235ee32bc0d10f4b7ec8a718f3d24a53660f679c

Could you elaborate on why "it won't work"? You can see the source that populated the GPF via the .nupkg.metadata file. 

> We don't want to display README in the browse tab if they aren't available from the remote source.
Great thing to explain the spec. Or if it's already there, please give me a pointer to that.

**Review by jebriede — APPROVED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageFileService.cs (hunk 12 above)
Comment (jgonz120) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageFileService.cs:172]: To clarify the requirement. We only want to display README from local storage in the installed tab. That way if there isn't a remote source setup with README exposed, they won't have as much of a inconsistent README experience in the browse tab. 

So I feel like this solution won't work because it's leaving it up to the PackageFileService to determine where to pull the README from, but it doesn't have the context if we're viewing the details from the Browse or Installed tabs

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Services/NuGetPackageFileServiceTests.cs (hunk 13 above)
Comment (jgonz120) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Services/NuGetPackageFileServiceTests.cs]: whoops, fixed

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageFileService.cs (hunk 12 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageFileService.cs:172]: I'm not seeing this explicitly in the Spec. I suggest updating that. I'd like to discuss the design, and the spec would be the best place for that.
